### ISSUE TITLE
hash mismatch on reading bloomfilter

### ISSUE THREAD

**Issue by thomasjungblut** [steakknife/bloomfilter#7]
hey!

I receive a strange error on reading a bloomfilter I just wrote: `Hash mismatch, the bloomfilter is probably corrupt.`

I turned on the debug output and got this back:

```
2018/07/25 20:35:48 New optimal bloom filter :: requested max elements (n):1000, probability of collision (p):0.0100000000 -> recommends -> bits (m): 4606 (0.000001 GiB), number of keys (k): 4
2018/07/25 20:35:48 bloomfilter.MarshalBinary: Successfully wrote [4 0 0 0 0 0 0 0 7 0 0 0 0 0 0 0 254 17 0 0 0 0 0 0 108 177 8 49 127 112 15 126 126 247 228 45 238 106 74 208 58 29 93 195 64 96 96 222 67 118 156 210 226 245 33 178 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 247 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 253 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 252 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 155 5 168 65 106 55 174 68 35 67 165 147 184 148 58 207 171 157 184 36 4 170 67 238 214 47 128 146 158 128 145 142 151 63 221 237 54 60 93 34 152 245 54 17 128 142 39 23] byte(s), sha384 [155 5 168 65 106 55 174 68 35 67 165 147 184 148 58 207 171 157 184 36 4 170 67 238 214 47 128 146 158 128 145 142 151 63 221 237 54 60 93 34 152 245 54 17 128 142 39 23]
2018/07/25 20:35:48 read bf k=4 n=7 m=4606
2018/07/25 20:35:48 bloomfilter.UnmarshalBinary() sha384 hash failed: actual [155 5 168 65 106 55 174 68 35 67 165 147 184 148 58 207 171 157 184 36 4 170 67 238 214 47 128 146 158 128 145 142 151 63 221 237 54 60 93 34 152 245 54 17 128 142 39 23]  expected [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
```

As you can see the actual hash is calculated correctly and also was written to the file previously.  However the expected is all zeros and that's causing a mismatch.

Here's a hexdump of the unzipped version of the file, I can send you the file as well if that helps you.
```
04 20 20 20 20 20 20 20 07 20 20 20 20 20 20 20 fe 11 20 20 20 20 20 20 05 3a 43 5a 5b 0d e7 0a 0d 73 0a 27 83 f4 34 60 a0 8c b5 e9 42 3d 6c 49 d6 d8 26 8b bf ae c1 97 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 7c 03 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 f8 03 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 f7 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 e0 0f 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 70 45 6a fb 39 a3 f2 66 73 c8 04 d0 4a 47 cb a4 33 b8 dd dc 37 6a 4f 5f 79 9d 8b b7 c5 c8 0b a1 6d 80 1c 44 93 4e 80 fc e3 e6 f3 47 0b c6 49 89  
```

I'm running go version go1.8.3 windows/amd64 and version 1.0.2 of this project.

---

Stepping through it with a debugger, there seems to be a mismatch in "bits" written (72) and read (n=7). It then just proceeds to read a bunch of zeros in the middle. Any idea where this inconsistency might come from?

**Comment by steakknife** [comment]
Fixed in cbf8825. The written format should be stable. It was a bug in the unmarshaller. Thanks to @thomasjungblut to reporting and clarifying the issue.

**Comment by steakknife** [comment]
Please comment and I'll reopen if there's anything else wrong with it.
